### PR TITLE
Saving GPU memory during pre-loading

### PR DESCRIPTION
**Pull request by wangyida**
### Problem: 
loading all training images into **pytorch tensor** is not necessary, which may lead to **OOM error** when training data is with a large scale.

### Solution: 
This PR makes it possible to further lower ```img_downscale``` ratio at least twice then the initial limit by loading all images into **numpy array** before batched training for COLMAP data format.
By replacing COLMAP dataloader from loading images into **tensors** to **numpy array**, the GPU memory could be saved a lot, especially when your training dataset is with a high resolution, i.e. 2000 x 2000, or with a large amount, i.e. > 1000 images.

Such changed dataloader can perform a high resolution training with certain data (1000 images - 2000 x 2000 pixels), which is not doable in the original codes.

The training speed is not influenced from my opinion and the accuracy is the same from mipnerf360 data.

- [x] test for mipnerf360 
- [x] nerf backend
- [x] neus backend
- [x] speed/psnr test - differs less than 5%
- [x] feedback from host @bennyguo

### CODE REVIEW

**Comment by bennyguo** [comment]
Hi @wangyida ! It makes sense as loading all images to gpu will consume large memory, but it'll also bring additional data transfer budget. I think it's better to add a `load_data_on_gpu` option to the dataset and let the user choose whether to transfer data on the fly or not. Could you add this option to the dataset and the corresponding config files? Besides, I would prefer to load them as torch CPU tensors instead of numpy arrays to keep a consistent style between the two options. Would you please modify this too? Thanks for your contribution!

**Comment by wangyida** [comment]
Sure I can modify it accordingly as you have suggested, thanks for sharing your nice work by the way. 

> Hi @wangyida ! It makes sense as loading all images to gpu will consume large memory, but it'll also bring additional data transfer budget. I think it's better to add a `load_data_on_gpu` option to the dataset and let the user choose whether to transfer data on the fly or not. Could you add this option to the dataset and the corresponding config files? Besides, I would prefer to load them as torch CPU tensors instead of numpy arrays to keep a consistent style between the two options. Would you please modify this too? Thanks for your contribution!

**Comment by wangyida** [comment]
Can you review the changes again @bennyguo , the system structures are kept unchanged now, just few lines changes in the dataloader

File: systems/nerf.py [hunk 1]
```diff
@@ -31,6 +32,7 @@ def forward(self, batch):
         return self.model(batch['rays'])
     
     def preprocess_data(self, batch, stage):
+        self.rank = get_rank()
```
Comment (bennyguo) [inline systems/nerf.py:35]: `self.rank` has already been set in `BaseSystem` so this is not needed:
https://github.com/bennyguo/instant-nsr-pl/blob/1f2ef5b0debfc095304afb8a973eb0aa884d3c02/systems/base.py#L18

File: systems/neus.py [hunk 2]
```diff
@@ -32,6 +33,7 @@ def forward(self, batch):
         return self.model(batch['rays'])
     
     def preprocess_data(self, batch, stage):
+        self.rank = get_rank()
```
Comment (bennyguo) [inline systems/neus.py:36]: save as above

File: datasets/colmap.py [hunk 3]
```diff
@@ -189,6 +189,7 @@ def setup(self, config, split):
                     img = Image.open(img_path)
                     img = img.resize(img_wh, Image.BICUBIC)
                     img = TF.to_tensor(img).permute(1, 2, 0)[...,:3]
+                    img = img.cuda() if self.config.load_data_on_gpu else img.cpu()
```
Comment (bennyguo) [inline datasets/colmap.py:192]: could you please change `img.cuda()` to `img.to(self.rank)` so that this works on DDP?

**Review by bennyguo — COMMENTED** [review]
Looks good! There're several minor issues that I've pointed out in the comments. After fixing these I think this PR is ready to be merged. Cheers!

**Comment by wangyida** [comment]
> 

Done thanks

**Comment by bennyguo** [comment]
Great! Thanks:)
